pkg/controller/nodeipam/ipam/multicidrset/multi_cidr_set.go

```diff
@@ -0,0 +1,365 @@
+		CIDR: s.Label,
+	}
+}
+
+func (s *MultiCIDRSet) getBeginningAndEndIndices(cidr *net.IPNet) (int, int, error) {
+	if cidr == nil {
+		return -1, -1, fmt.Errorf("error getting indices for cluster cidr %v, cidr is nil", s.ClusterCIDR)
+	}
+	begin, end := 0, s.MaxCIDRs-1
+	cidrMask := cidr.Mask
+	maskSize, _ := cidrMask.Size()
+	var ipSize int
+
+	if !s.ClusterCIDR.Contains(cidr.IP.Mask(s.ClusterCIDR.Mask)) && !cidr.Contains(s.ClusterCIDR.IP.Mask(cidr.Mask)) {
+		return -1, -1, fmt.Errorf("cidr %v is out the range of cluster cidr %v", cidr, s.ClusterCIDR)
+	}
+
+	if s.clusterMaskSize < maskSize {
+		var err error
+		ipSize = net.IPv4len
+		if cidr.IP.To4() == nil {
+			ipSize = net.IPv6len
+		}
+		begin, err = s.getIndexForCIDR(&net.IPNet{
+			IP:   cidr.IP.Mask(s.nodeMask),
+			Mask: s.nodeMask,
+		})
+		if err != nil {
+			return -1, -1, err
+		}
+		ip := make([]byte, ipSize)
+		if cidr.IP.To4() != nil {
+			ipInt := binary.BigEndian.Uint32(cidr.IP) | (^binary.BigEndian.Uint32(cidr.Mask))
+			binary.BigEndian.PutUint32(ip, ipInt)
+		} else {
+			// ipIntLeft          |         ipIntRight
+			// 2001:0DB8:1234:0000:0000:0000:0000:0000
+			ipIntLeft := binary.BigEndian.Uint64(cidr.IP[:net.IPv6len/2]) | (^binary.BigEndian.Uint64(cidr.Mask[:net.IPv6len/2]))
+			ipIntRight := binary.BigEndian.Uint64(cidr.IP[net.IPv6len/2:]) | (^binary.BigEndian.Uint64(cidr.Mask[net.IPv6len/2:]))
+			binary.BigEndian.PutUint64(ip[:net.IPv6len/2], ipIntLeft)
+			binary.BigEndian.PutUint64(ip[net.IPv6len/2:], ipIntRight)
+		}
+		end, err = s.getIndexForCIDR(&net.IPNet{
+			IP:   net.IP(ip).Mask(s.nodeMask),
+			Mask: s.nodeMask,
+		})
+		if err != nil {
+			return -1, -1, err
+		}
+	}
+	return begin, end, nil
+}
+
+// Release releases the given CIDR range.
+func (s *MultiCIDRSet) Release(cidr *net.IPNet) error {
+	begin, end, err := s.getBeginningAndEndIndices(cidr)
+	if err != nil {
+		return err
+	}
+	s.Lock()
+	defer s.Unlock()
+
+	for i := begin; i <= end; i++ {
+		// Remove from the allocated CIDR Map and decrement the counter only if currently
+		// marked allocated. Avoids double counting.
+		currCIDR, err := s.indexToCIDRBlock(i)
+		if err != nil {
+			return err
+		}
+		if _, ok := s.AllocatedCIDRMap[currCIDR.String()]; ok {
+			delete(s.AllocatedCIDRMap, currCIDR.String())
+			s.allocatedCIDRs--
+			cidrSetReleases.WithLabelValues(s.Label).Inc()
+		}
+	}
+
+	cidrSetUsage.WithLabelValues(s.Label).Set(float64(s.allocatedCIDRs) / float64(s.MaxCIDRs))
+
+	return nil
+}
+
+// Occupy marks the given CIDR range as used. Occupy succeeds even if the CIDR
+// range was previously used.
+func (s *MultiCIDRSet) Occupy(cidr *net.IPNet) (err error) {
+	begin, end, err := s.getBeginningAndEndIndices(cidr)
+	if err != nil {
+		return err
+	}
+	s.Lock()
+	defer s.Unlock()
+
+	for i := begin; i <= end; i++ {
+		// Add to the allocated CIDR Map and increment the counter only if not already
+		// marked allocated. Avoids double counting.
```

Comment:
Is double counting expected behavior? If not, do we need to log an error?

Reply:
No not expected, modified the comment to say `Prevents double counting` instead of `Avoids double counting`. As we are preventing it, no error needs to be logged